### PR TITLE
Bug 1391702 - Replace Bugzilla::User::validate_password() with calls to Data::Password::passwdqc

### PR DESCRIPTION
**Pull request by dylanwh**

### CODE REVIEW

File: Bugzilla/Config/Auth.pm [hunk 1]
```diff
@@ -107,18 +107,41 @@ sub get_param_list {
...
+        {
+            name => 'passwdqc_match_length',
+            type => 't',
+            default => '4',
+        },
+
+        {
+            name => 'passwdqc_random_bits',
+            type => 't',
+            default => '47',
```
Comment (globau) [inline Bugzilla/Config/Auth.pm]: these should use the `check_numeric` checker

File: Bugzilla/Config/Auth.pm [hunk 2]
```diff
@@ -107,18 +107,41 @@ sub get_param_list {
...
-        {
             name    => 'password_check_on_login',
             type    => 'b',
             default => '1'
         },
+
+        {
+            name => 'passwdqc_min',
+            type => 't',
+            default => 'undef, 24, 11, 8, 7',
```
Comment (globau) [inline Bugzilla/Config/Auth.pm]: this needs a checker to ensure the correct number of elements are provided, and that each element is either "undef" or an int.

File: template/en/default/admin/params/auth.html.tmpl [hunk 3]
```diff
@@ -24,6 +24,79 @@
...
+
+[%  desc_passwdqc_max = BLOCK %]
+  The maximum allowed password length. This can be used to prevent users
+  from setting passwords that may be too long for some system services.
+  <p>
+  The value 8 is treated specially: with "max=8", passwords longer than
+  8 characters will not be rejected, but will be truncated to 8
+  characters for the strength checks and the user will be warned. This
+  is to be used with the traditional DES-based password hashes, which
+  truncate the password at 8 characters.
```
Comment (globau) [inline template/en/default/admin/params/auth.html.tmpl]: this paragraph doesn't seem relevant to bugzilla, since it doesn't use oldschool hashes.  is it possible to disable the special handling of `max=8`?

File: Bugzilla/Install.pm [hunk 4]
```diff
@@ -500,9 +500,14 @@ sub _prompt_for_password {
...
         my $pass2 = <STDIN>;
         chomp $pass2;
-        eval { validate_password($password, $pass2); };
-        if ($@) {
-            print "\n$@\n";
+        my $pwqc = Bugzilla->passwdqc;
+        my $ok = $pwqc->validate_password($password);
+        if (!$ok) {
+            print "\n", $pwqc->reason, "\n";
+            undef $password;
```
Comment (globau) [inline Bugzilla/Install.pm]: it may be worth documenting the default password complexity requirements in the README, to ease the initial setup.

File: Bugzilla/WebService/Constants.pm [hunk 5]
```diff
@@ -159,7 +158,6 @@ use constant WS_ERROR_CODE => {
     auth_cant_create_account    => 501,
     account_creation_disabled   => 501,
     account_creation_restricted => 501,
-    password_too_short    => 502,
```
Comment (globau) [inline Bugzilla/WebService/Constants.pm]: akin to 503, add a comment that 502 password_too_short no longer exists

File: qa/t/test_bmo_enter_new_bug.t [hunk 6]
```diff
@@ -242,6 +242,7 @@ $sel->type_ok("comment", "--- Bug created by Selenium ---", "Enter bug descripti
 $sel->click_ok("commit", undef, "Submit bug data to post_bug.cgi");
 $sel->wait_for_page_to_load_ok(WAIT_TIME);
 $sel->is_text_present_ok('has been added to the database', 'Bug created');
+screenshot_page($sel, "/app/artifacts/power_bug_id.png");
```
Comment (globau) [inline qa/t/test_bmo_enter_new_bug.t]: doesn't appear to be related to this change.

File: reset_password.cgi [hunk 7]
```diff
@@ -49,7 +48,14 @@ if ($cgi->param('do_save')) {
...
     }
-    validate_password($password_1, $password_2);
+
+    my $pwqc = Bugzilla->passwdqc;
+    if (not $pwqc->validate_password($password_1)) {
+        ThrowUserError("password_insecure", { reason => $pwqc->reason });
+    }
+    elsif ($password_1 ne $password_2) {
+        ThrowUserError("password_mismatch");
+    }
```
Comment (globau) [inline reset_password.cgi]: you should be checking that `$password_1 eq $password_2` _before_ checking the password complexity.

this same logic is repeated multiple times, so either apply the same fix everywhere, or move `password_mismatch` checking into a common fn.

File: template/en/default/admin/params/auth.html.tmpl [hunk 8]
```diff
@@ -24,6 +24,79 @@
...
+  partially discounted.
+  <p>
+  The substring search is case-insensitive and is able to detect and
+  remove a common substring spelled backwards.
+[% END %]
+
+[%  desc_random_bits = BLOCK %]
+  The size of randomly-generated passphrases in bits (24 to 85), or 0 to
+  disable this feature.
+[% END %]
```
Comment (globau) [inline template/en/default/admin/params/auth.html.tmpl]: i don't think this should be an admin setting, as `generate_password` is only called by `generate_test_data.pl`.  it isn't clear what disabling this feature would do.

File: Bugzilla/Config/Auth.pm [hunk 9]
```diff
@@ -107,18 +107,41 @@ sub get_param_list {
         },
 
         {
-            name    => 'password_complexity',
```
Comment (globau) [inline Bugzilla/Config/Auth.pm]: `password_complexity` is still used by `account/email/confirm-new.html.tmpl` and `global/password-features.html.tmpl`.

**Comment by dylanwh** [comment]
Removing the UI is outside the scope of this bug, that's in bug 1390903 /  https://github.com/mozilla-bteam/bmo/pull/205
Both of these will land at the same time.

File: reset_password.cgi (hunk 7 above)
Comment (dylanwh) [inline reset_password.cgi]: No, it is intentional to check the password sameness second, as passwdqc also lets us set a max password. It is useful to avoid comparing huge strings.

File: Bugzilla/Config/Auth.pm (hunk 9 above)
Comment (dylanwh) [inline Bugzilla/Config/Auth.pm]: I'll skip removing that in this patch and move it to #205

File: template/en/default/admin/params/auth.html.tmpl (hunk 8 above)
Comment (dylanwh) [inline template/en/default/admin/params/auth.html.tmpl]: we'll be exposing generate_password() in a later patch, but random_bits will be always 85.

File: qa/t/test_bmo_enter_new_bug.t (hunk 6 above)
Comment (dylanwh) [inline qa/t/test_bmo_enter_new_bug.t]: debugging screenshot. :-)

**Comment by dylanwh** [comment]
So almost ready for another review. IMHO repeating the password check code is preferable to having some opaque function like "valid_password" that sometimes takes one argument, sometimes two, and sometimes returns a boolean and at other times throws an exception (actually, there were two functions on that axis, but I digress)

Anyway, I will factor out the password validation *and* "did you type it correctly?" logic into a *verbosely* named function exported from Bugzilla::Util, something like:

```perl
sub assert_password_is_secure_and_typed_correctly {
my $pwqc = Bugzilla->passwdqc;
    if (not $pwqc->validate_password($pwd1)) {
        ThrowUserError('password_insecure', { reason => $pwqc->reason });
    }
    elsif ($pwd1 ne $pwd2) {
        ThrowUserError('password_mismatch');
    }
}
```

**Comment by floatingatoll** [comment]
`if ($pwqc->insecure($pwd1)) { ThrowUserError('password_insecure'` seems like it would be simpler to read in code, fwiw.

**Comment by dylanwh** [comment]
yes, but I'm also not about to change the API of a dependency. :-)

**Comment by floatingatoll** [comment]
👍

**Comment by dylanwh** [comment]
Note: leave-open until after tomorrow's push.

**Comment by dylanwh** [comment]
I've added code here that changes the way we handle insecure passwords on login. Discussion to follow on the bug.

**Comment by dylanwh** [comment]
looks like kang approved of this last change -- so @globau can re-approve?
